src/test/java/com/faforever/api/deployment/ExeUploaderServiceTest.java

```diff
@@ -0,0 +1,125 @@
+    bytes = new byte[]{1, 2, 3, 4};
+    when(contentService.createTempDir()).thenReturn(temporaryDirectory.getRoot().toPath());
+
+    featuredModFile = new FeaturedModFile();
+    featuredModFile.setName("ForgedAlliance.exe");
+    featuredModFile.setVersion(1);
+    featuredModFile.setFileId((short) 1);
+  }
+
+  @After
+  public void after() {
+    verifyNoMoreInteractions(contentService, apiProperties, deployment, featuredModService);
+  }
+
+  @Test
+  public void testProcessUploadBeta() {
+    String modName = "fafbeta";
+
+    String finalExeDestination = finalDirectory.getRoot().getAbsolutePath() + "/ForgedAlliance.1.exe";
+    when(deployment.getForgedAllianceBetaExePath()).thenReturn(finalExeDestination);
+
+    when(featuredModService.getFile(modName, null, "ForgedAlliance.exe")).thenReturn(
+      featuredModFile);
+    instance.processUpload(bytes, modName);
+
+    // File was created
+    assertTrue(Files.exists(Paths.get(finalExeDestination)));
+    // File was updated in database
+    ArgumentCaptor<List<FeaturedModFile>> modFilesCaptor = ArgumentCaptor.forClass(List.class);
+    verify(featuredModService).save(eq(modName), eq((short) featuredModFile.getVersion()), modFilesCaptor.capture());
+    assertThat(modFilesCaptor.getValue().size(), is(1));
+    assertThat(modFilesCaptor.getValue(), hasItem(featuredModFile));
+
+    verify(contentService).createTempDir();
+    verify(apiProperties, atLeastOnce()).getDeployment();
+    verify(deployment).getForgedAllianceBetaExePath();
+    verify(featuredModService).getFile(modName, null, "ForgedAlliance.exe");
+  }
+
+  @Test
+  public void testProcessUploadDevelop() {
+    String modName = "fafdevelop";
+
+    String finalExeDestination = finalDirectory.getRoot().getAbsolutePath() + "/ForgedAlliance.1.exe";
+    when(deployment.getForgedAllianceDevelopExePath()).thenReturn(finalExeDestination);
+
+    when(featuredModService.getFile(modName, null, "ForgedAlliance.exe")).thenReturn(
+      featuredModFile);
+    instance.processUpload(bytes, modName);
+
+    // File was created
+    assertTrue(Files.exists(Paths.get(finalExeDestination)));
+    // File was updated in database
+    ArgumentCaptor<List<FeaturedModFile>> modFilesCaptor = ArgumentCaptor.forClass(List.class);
+    verify(featuredModService).save(eq(modName), eq((short) featuredModFile.getVersion()), modFilesCaptor.capture());
+    assertThat(modFilesCaptor.getValue().size(), is(1));
+    assertThat(modFilesCaptor.getValue(), hasItem(featuredModFile));
+
+    verify(contentService).createTempDir();
+    verify(apiProperties, atLeastOnce()).getDeployment();
+    verify(deployment).getForgedAllianceBetaExePath();
+    verify(deployment).getForgedAllianceDevelopExePath();
+    verify(featuredModService).getFile(modName, null, "ForgedAlliance.exe");
+  }
+
+  @Test(expected = ApiException.class)
+  public void testProcessUploadIsForbidden() {
+    String modName = "faf";
```

Comment:
Not sure if that is already possible in the junt version used here but if so use assertThrows and specify more clearly what the Exception looks like

Reply:
Needs JUnit5, can be refactored once Map refactor is merged.